Signup: refactor search results logic

ramonjd opened the pull request:
⚠️ This PR is a little on the large side, so if it works okay I can split it up :)

## Changes proposed in this Pull Request

This is an attempt to incorporate many of the suggestions we received in our spike PR #31378, including, but not limited to:

### ~~Debouncing the site preview update so that we don't get 'flashes' of the default site preview as the user types~~

_Update_: We will address this in ~~a future PR~~ https://github.com/Automattic/wp-calypso/pull/32055. See comment: https://github.com/Automattic/wp-calypso/pull/31708#issuecomment-480145319

![debounce](https://user-images.githubusercontent.com/6458278/55216081-ed619480-524f-11e9-9adb-6cc0fd81186b.gif)

### Caching previous results, which we can display to the user if their query yields none (probably should be migrated to the backend eventually)
![no-match](https://user-images.githubusercontent.com/6458278/55216120-05391880-5250-11e9-8a32-f42cd5bd72c2.gif)

### Removing sorting on the clientside
No GIF 🕶 

## Testing instructions
Fire up the branch and search for a vertical. In particular:

* Select a popular category, clear the field, select another
* Enter text, slow and fast
* _More scenarios to come..._

ramonjd committed:
Updating search logic:
- Ensuring that already-fetched results persist when the user has searched for a non-result yielding query
- Perform vertical matches on exact and partial matches

Adding results to the internal state

Remove sortResults as the API does its own sorting now.

Removing deprecated utils.js

creating a custom callback for popular topics

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -83,98 +144,72 @@ export class SiteVerticalsSuggestionSearch extends Component {
+			value !== this.props.searchValue &&
+			size( value ) >= this.props.charsToTriggerSearch
 		) {
 			this.setState( { railcar: this.getNewRailcar() } );
```
@adamziel 

Since we're not using this condition block, what would be the conditions for railcar now?

southp replied:
My two cents for `value && value !== this.props.searchValue` since it seems that we can deprecate `chartsToTriggerSearch` . If this sounds good, I think we can also remove that TODO comment. I wrote that because I thought it wasn't the best place, but it now looks about right after your work here :)

southp commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -83,98 +144,72 @@ export class SiteVerticalsSuggestionSearch extends Component {
+				<QueryVerticals
+					searchTerm={ this.state.searchValue.trim() }
+					debounceTime={
+						! this.state.isSuggestionSelected || size( this.state.searchValue ) > 4 ? 300 : 0
```
I personally don't quite understand the purpose behind this condition. Why do we want to alter the debouncing time based on this? How does it reflect on the UX?

ramonjd replied:
Me either. I think it was an oversight, that is, I left it in there after test. 

Thanks for catching this (again) :)

southp commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -52,9 +50,61 @@ export class SiteVerticalsSuggestionSearch extends Component {
+	 *
+	 * This function could better be performed in the backend eventually.
+	 *
+	 * @param {Array} results Incoming HTTP results
```
One major purpose of the existence of the data layer, redux thunks or any other effectful data request paradigm is to isolate app state and UI components from the data source. In this case, `setSearchResults()` shouldn't be aware of the HTTP thing, but only the data shape of `results`. Thus, let's get rid of all the HTTP-related things in the description :)

southp commented on client/components/site-verticals-suggestion-search/utils.js:
```diff
@@ -0,0 +1,59 @@
+/** @format */
```
I think this whole module be deprecated now?

ramonjd replied:
Ah good call!

southp reviewed (dismissed):
The test plan works for me. However, it's quite hard for me to reason the purposes and how each line of change maps to them. If not too much burden, I'd suggest to segregate this PR so that we can have a deeper look into each ingredient and have a better guarantee that nothing unexpected got slipped in.

ramonjd commented on client/components/site-verticals-suggestion-search/README.md:
```diff
@@ -26,7 +26,7 @@ render() {
 
-### _(String)_ `initialValue`
+### _(String)_ `searchValue`
 An _optional_ initial value of the search input field. Default is `''`;
```
I think `"The search field value."` should be enough for now :)

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -63,18 +113,29 @@ export class SiteVerticalsSuggestionSearch extends Component {
 
-	// When a user is keying through the results,
-	// only update the vertical when they select a result.
+	/***
```
I'm clearly starstruck 🌟  :)

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -52,9 +50,61 @@ export class SiteVerticalsSuggestionSearch extends Component {
+				// set a flag to load sans debounce
+				isSuggestionSelected
+					? this.updateVerticalData( matches, this.state.searchValue )
+					: this.updateVerticalDataDebounced( matches, this.state.searchValue )
```
> why we would need to debounce state updating here

Ooh, that was a bad rebase. You're right @southp . It doesn't need to be there. Nice 👀 !!

But to avoid updating the state too frequently we need it `[onChange](https://github.com/Automattic/wp-calypso/pull/31708/files/d64a4faf7499774bb41d009b7c077a7c61c11526#diff-9553ff7100a8a69d6011c3bc6f8dded2R177)` at least.

**With no debounce:**
![no-debounce-onchange](https://user-images.githubusercontent.com/6458278/55372464-8ee92e80-554d-11e9-8bc8-f0953aa1783a.gif)

**With:**
![debounce-on-change](https://user-images.githubusercontent.com/6458278/55372504-b0e2b100-554d-11e9-8ac9-5ea8a8d88fab.gif)

I'll look more closely at this today.

tyxla commented on client/state/signup/verticals/reducer.js:
```diff
@@ -4,14 +4,22 @@
+		[ SIGNUP_VERTICALS_REQUEST ]: state => {
+			return {
+				...state,
+				isFetching: true,
```
I usually advise against any `isFetching`/`isRequesting` when possible. If you're sure we need it and there's no other way - that's fine.

ramonjd replied:
Thanks for raising this. Normally I'd have not added this to the state. There was some headscratching around the problem.

I should be able to wrangle it, but I'll need to add a custom prop method, something like:

```
isVerticalSearchPending: ( searchValue, verticals = [] ) => searchValue && 0 === verticals.length,
```

Or something... :)

tyxla commented on client/state/signup/verticals/reducer.js:
```diff
@@ -4,14 +4,22 @@
+		[ SIGNUP_VERTICALS_REQUEST ]: state => {
+			return {
+				...state,
+				isFetching: true,
```
This feels like an unexpected location in the tree for storing a `isFetching` variable. I expected it to be in a separate subtree of the state.

ramonjd replied:
Totally agree. Will fix this.

tyxla commented on client/state/signup/verticals/test/reducer.js:
```diff
@@ -23,6 +38,7 @@ describe( 'state/signup/verticals/reducer', () => {
 			} )
 		).toEqual( {
 			foo: verticals,
+			isFetching: false,
```
Is that worth a separate test?

tyxla reviewed (commented):
This tests great for the Jetpack use case (`/jetpack/connect/site-topic/:site`) 👍 

Left some questions and comments, mostly minor stuff. Nice work!

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -83,98 +139,62 @@ export class SiteVerticalsSuggestionSearch extends Component {
-			valueLengthShouldTriggerSearch &&
-			// Don't trigger a search if there's already an exact, non-user-defined match from the API
-			! result
+			!! value &&
```
👍 Good spot.

I hope we can get rid of this condition block all together once the regily question is cleared up.

ramonjd commented on client/components/site-verticals-suggestion-search/README.md:
```diff
@@ -26,8 +26,8 @@ render() {
 
 ## Props
 
-### _(String)_ `initialValue`
```
Nothing to see here! :trollface: 

I missed that. Thanks for the 👀

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -31,7 +31,6 @@ export class SiteVerticalsSuggestionSearch extends Component {
 		onChange: PropTypes.func,
 		placeholder: PropTypes.string,
 		shouldShowPopularTopics: PropTypes.func,
-		searchValue: PropTypes.string,
```
I've added `searchValue`. `initialValue` was the old prop name. It just doesn't want to die. :)

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -52,9 +50,56 @@ export class SiteVerticalsSuggestionSearch extends Component {
+			if (
+				! isSuggestionSelected &&
+				! find( results, item => ! item.isUserInputVertical ) &&
+				1 < size( candidateVerticals )
```
Nope. Removing. 👍

ramonjd commented:
Thanks a lot for the helpful review @tyxla and @southp !!!

I've updated according to your recommendations. 🤞

southp commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -26,35 +26,82 @@ import './style.scss';
+			if (
+				! isSuggestionSelected &&
+				! find( results, item => ! item.isUserInputVertical ) &&
+				1 < candidateVerticals.length
```
FYI no need to use Yoda condition in calypso, but please feel free to keep it if it pleases you :)

ramonjd replied:
I know, it's a habit that's hard to break now, like ( spaces ) in all my ( parentheses ) 😆

southp commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -26,35 +26,82 @@ import './style.scss';
 			railcar: this.getNewRailcar(),
 			candidateVerticals: [],
+			isSuggestionSelected: false,
+			searchValue: props.searchValue,
```
I'm going to touch one of the hardest CS problem here: Could we name these two state / props differently so we can distinguish them better?

I personally found a few times that I misunderstood the code because I misread `this.state.searchValue` as `this.props.searchValue`, and vice versa. Thus, it would be great if we can have better naming here.

I feel `searchValue` for the state is quite nice, but what would match the meaning of the current `this.props.searchValue` better? In my previous run of refactoring, it becomes a props because the component doesn't hold the input box value itself anymore. Now that the react state one is back with a legit reason (I think it's because debouncing. Please feel free to correct me if I'm wrong),  it's time to revisit the naming and the necessity of this prop.

I don't really know the best direction of this, so this is more of a brain dump comment than a request to fix. Please feel free to keep it this way and we can revisit it in the future.

ramonjd replied:
I think it was a product of the rebase after the state tree updates.

I'll change it back to my original name `initialValue`, which describes the prop more accurately now since we're using the internal state to track updates.

Cheers!

southp reviewed (commented):
@ramonjd Thanks for the helpful clarification. I can now understand the whole picture of this PR. With  the in-depth look of @tyxla , there is also definitely no need to split anymore :)

It works pretty well for me, both on the onboarding flow and the main flow. I've noted a few more places that I missed previously and would worth polishing. Please feel free to DM me if you want a timely revisit 🙇🏻

tyxla commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -26,35 +26,82 @@ import './style.scss';
 
 export class SiteVerticalsSuggestionSearch extends Component {
 	static propTypes = {
+		autoFocus: PropTypes.bool,
```
We don't seen to be documenting `autoFocus` in the README - now is a good chance to do that.

tyxla reviewed (commented):
I think this is getting closer 👍 

I agree with @southp's feedback, in which are brought some really good points.

Haven't re-tested, but will do on the next round of reviews. Left some more comments in the meantime.

ramonjd commented:
Great feedback @southp and @tyxla 

Looking at this today. 

Happy that we can address these things now, and not after merge! 

Getting closer... :)

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -184,15 +213,16 @@ export default localize(
+			bindDebounce: func => debounce( func, 1000 ),
 			shouldShowPopularTopics: searchValue => ! searchValue && ownProps.showPopular,
+			isVerticalSearchPending: ( searchValue, verticals = [] ) =>
+				searchValue && 0 === verticals.length,
```
> If we have to mock them in unit tests, define them through defaultProps could work.

That was the main motivation, particularly for the `bindDebounce`. We could also extract them into a `utils.js` and mock the import methods.

👍

ramonjd replied:
Or just mock `debounce` 🎉

ramonjd commented on client/components/site-verticals-suggestion-search/popular-topics.jsx:
```diff
@@ -65,7 +65,7 @@ class PopularTopics extends PureComponent {
 		event.preventDefault();
 		event.stopPropagation();
-		this.props.onSelect( event.currentTarget.value );
+		this.props.onSelect( event.currentTarget.value, true );
```
> If I understand correctly, the reason that we are passing true here is to comply with the parameter list of onChange() of <SuggestionSearch/>

I was trying to conform to the `isNavigating` bool. I can extract that as you suggested and resolve it in the consuming component. 🙇

ramonjd commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -184,15 +213,16 @@ export default localize(
+			bindDebounce: func => debounce( func, 1000 ),
 			shouldShowPopularTopics: searchValue => ! searchValue && ownProps.showPopular,
+			isVerticalSearchPending: ( searchValue, verticals = [] ) =>
+				searchValue && 0 === verticals.length,
```
Hmmm, anything is possible. If it can't find a matching vertical, the API will always return one result, that is, whatever the user has inputted. 

But it won't hurt to be defensive here.

👍

ramonjd replied:
Now that I think of it, I think that was one of the reasons why I added `isFetching` to the state 🤣

tyxla commented on client/components/site-verticals-suggestion-search/README.md:
```diff
@@ -15,9 +15,9 @@ onChange( { vertical_name, vertical_slug, vertical_id } ) {
-			onChange={ this.onChange }
-			initialValue={ this.state.initialValue }
-			charsToTriggerSearch={ 3 }
+			autoFocus={ true }
```
Nit: `={ true }` isn't necessary; we could just have `autoFocus` and it would do the same thing.

ramonjd replied:
Yeah, I'm being intentionally explicit here. 👍

tyxla commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -26,55 +26,126 @@ import './style.scss';
+	 * Searches the API results for a direct match on the user search query.
+	 *
+	 * @param {String} value Search query array
+	 * @returns {Object?} An object from the vertical results array
```
Docs nit: unless I'm missing something, this function will return `undefined` if there are no matches, and not `null`. We should reflect that in the docs (removing the `?`).

In case I'm right about this, let's make sure any usages of `searchForVerticalMatches` work well with the `undefined` case.

tyxla commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -83,116 +154,74 @@ export class SiteVerticalsSuggestionSearch extends Component {
+				isSuggestionSelected,
+				inputValue: value,
+			},
+			() => this.updateVerticalDataDebounced( this.searchForVerticalMatches( value ), value )
```
Since we'll debounce this anyway, could we move this as a separate call after `this.setState()`? When the call is debounced for a second, there's little difference if we wait or not for the `setState()` to finish its job.

tyxla commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -26,55 +26,126 @@ import './style.scss';
-		shouldShowPopularTopics: PropTypes.func,
 		searchValue: PropTypes.string,
+		showPopular: PropTypes.bool,
+		shouldShowPopularTopics: PropTypes.func,
```
Seems like this one is no longer used in favor of the corresponding method.

tyxla commented on client/components/site-verticals-suggestion-search/README.md:
```diff
@@ -58,4 +52,9 @@ Returns _{Object}_:
+_Optional_ Informs the component whether to show a list of popular vertical topics when the input field is empty. Default: `false`
+
+### _(Boolean)_ `autoFocus` 
+_Optional_ When set to `true` gives immediate focus to the search input field. Default: `false`
```
Would be great if we maintain a thorough list of all props here. Definitely don't consider that as a requirement though - it's something that can be followed up with in another PR.

ramonjd replied:
I've taken the approach that the README is for the consumer of the component, and therefore lists the props required to embed it, not necessarily those defined via connect, which are only important for internals. 

Having said that, happy to add them later. I'm sure I'll be back in this neck of woods very soon! :)

tyxla commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -26,55 +26,126 @@ import './style.scss';
+	 * @param {String} value Search query array
+	 */
+	updateVerticalData = ( verticalData, value = '' ) => {
+		value = value.trim();
```
Oops. Let's avoid mutating the argument 😉 Creating a copy of it in another variable is fine.

tyxla commented on client/components/site-verticals-suggestion-search/index.jsx:
```diff
@@ -83,116 +154,74 @@ export class SiteVerticalsSuggestionSearch extends Component {
+	 * Callback to be passed to consuming component when the search field is updated.
+	 *
+	 * @param {String} value The new search value
+	 * @param {Bool} isSuggestionSelected Whether the user has selected a suggestion
```
Nit of the nits: I've found that vertically aligning the different fragments of the JSdoc makes it so much more readable:

```
	 * @param {String} value                The new search value
	 * @param {Bool}   isSuggestionSelected Whether the user has selected a suggestion
```

I'm fine either way tho - feel free to take this suggestion or throw it in the trash.

tyxla reviewed (commented):
Great work!

I assume this is still not ready for review, but I've left some more little feedback that could be helpful in the meantime. 90% is nits 😉 

Let me know when this is ready for another 👁

ramonjd committed:
Removing unused props and test comments
setting var for trimmed value
cleaning up jsdoc comments
calling setState callback `updateVerticalDataDebounced` directly after

ramonjd commented:
Thanks for helping make this patch better @tyxla and @southp 🙇 🙇 🙇 🙇 🙇 🙇 🙇

All suggestions implemented gladly :)

@southp  Any chance you might give it a whirl today for a possible launch?

ramonjd committed:
This commit removes the input debounce for now. The reason why we have to revisit this later is because we're tracking a user's selection `onSubmit`. This tracking event requires vertical data related to the user's input. Due to the debounce however, the field value is often added to the state too late. So we need to make sure we have the latest response that corresponds to the user input before the user can submit the form.
The main reason for the debounce was to prevent the site preview mockup from unsightly updates due to us hammering the state with updates. I think this indicates that we need to address this in the mockup rather than here.

ramonjd commented:
The latest commit removes the input debounce for now, to be revisited later.

The main reason for the debounce was to prevent the site preview mockup from unsightly 'flashes' due to hammering the state with updates.

We have to remove it however because we're tracking a user's selection `onSubmit`. 

This tracking event requires vertical data related to the user's input. 

Due to the debounce however, the field value is often added to the state too late. We need to make sure we have the latest response that corresponds to the user input before the user can submit the form.

Because of the input field debounce, we're not updating the state tree in time for `null == getVertical( state, searchValue )` to have any effect, that is, until the debounce is flushed, the field and the state are not synced. The result is that we send of stale data:

![Apr-05-2019 14-11-35](https://user-images.githubusercontent.com/6458278/55602544-e982c480-57b1-11e9-8a58-a80af747f867.gif)

I think this indicates that we need to address the mockup flashes probably in the mockup itself rather than in `<SiteVerticalsSuggestionSearch />`.

tyxla commented on client/signup/steps/site-topic/form.jsx:
```diff
@@ -101,8 +101,7 @@ class SiteTopicForm extends Component {
 		const siteTopic = getSiteVerticalName( state );
-		const isVerticalSearchPending = null == getVerticals( state, siteTopic );
-		const isButtonDisabled = ! siteTopic || isVerticalSearchPending;
+		const isButtonDisabled = ! siteTopic || null == getVerticals( state, siteTopic );
```
I assume the `null ==` here is intentionally not strict?

tyxla reviewed (approved):
Nice work 👍 Code looking good and working well for the Jetpack step.
